woocommerce/woocommerce
# @woocommerce/api: Added list() repository operation and classes for managing store settings

### All Submissions:






### Changes proposed in this Pull Request:

This PR adds support for repository `list()` operations, settings & setting groups, and adds type-safety to all of the repository methods.

#### `list()` Operation

This change is relatively small and just includes the addition of two new callbacks `ListFn` and `ListChildFn`, as well as two new interfaces, `ListsModels` and `ListsChildModels`.

#### Settings & Setting Groups

I've added these models in their entirety as well as list operations for both and fetch/update operations for settings. I then added a `SettingService` in order to make it easier to set common settings without having to find out their IDs.

#### Type-Safety

As I started to add the settings and groups I realized that there was a lot of boilerplate and that the type safety didn't always work as expected. In this PR I've introduced the following interface:

```typescript
interface ModelRepositoryParams<
	T extends Model = never,
	ParentID extends ModelParentID | undefined = never,
	ListParams = never,
	UpdateParams extends keyof T = never,
	> {
	thisTypeIsDeclarativeOnly: string;
}
```

This interface allows all of the repository parameters to be stored easily in a definition and then given to a repository. We're using conditional types to extract the generic parameters and then using them where necessary. While this may seem kind of complicated under the hood, the end result is that adding a new model and repository is even easier and the method signatures are a lot clearer.

#### Documentation Cleanup

I've gone through and cleaned up all of the types in the documentation as well as some of the descriptions in an effort to make it all easier to consume from a non-TypeScript project.

### How to test the changes in this Pull Request:

In addition to the `jest` tests included in the package, the E2E test suite consumes the simple product repository and so a success in the E2E suite means everything is still working. You can test the new setting service with this snippet placed in the `tests/e2e/api/src` directory, in a `service.test.ts` file:

```typescript
import { HTTPClientFactory } from './http';
import { Setting } from './models';
import { SettingService } from './services';

test( 'Test Setting Service', async () => {
	const client = HTTPClientFactory.withOAuth(
		'http://woocommerce.test/wp-json/',
		'consumer_key',
		'consumer_secret',
	);

	const service = new SettingService( Setting.restRepository( client ) );

	await service.updateStoreAddress(
		'1234 Fake Street',
		'Apt Test',
		'Testing',
		'US:WA',
		'12345',
	);
} );
```

You'll have to adjust the `hostname`, `consumer_key`, and `consumer_secret` to point at your local test site in order for this to work. Once you've done that you should be able to run `npm test` in the `tests/e2e/api` directory and it will succeed!

### Other information:
